### DISCUSSION TITLE
Code example for automatically executing repetitive task workflows from data input to report generation

### DISCUSSION THREAD

**Discussion by shxylgithub** [elsa-workflows/elsa-core#5648] [General]
Code example for automatically executing repetitive task workflows from data input to report generation

**Comment by glime-ai** [comment]
To automatically execute repetitive task workflows from data input to report generation using Elsa Workflows, you would typically create a workflow that includes activities for each step of the process. Below is a high-level example of how you might structure such a workflow programmatically:

```csharp
// Define workflow builder
var builder = new WorkflowBuilder();

builder
    .StartWith<ReceiveHttpRequest>(
        setup => setup
            .WithPath("/data-input")
            .WithMethod(HttpMethods.Post)
            .WithReadContent())
    .Then<ForEach>(
        // ForEach activity to process each item in the input data
        forEach => forEach
            .WithCollection(context => context.GetInput<IEnumerable<MyDataItem>>())
            .Do(x => x
                .StartWith<ProcessDataItem>()
                .Then<CustomActivity>() // Your custom activity to perform a repetitive task
            ))
    .Then<GenerateReport>() // Activity to generate a report
    .Then<SendHttpResponse>( // Send the generated report as a response
        setup => setup.WithResponse(context => context.GetVariable<Report>("Report"))
    );

// Build and start the workflow
var workflow = builder.Build();
var workflowInstance = await workflowInvoker.StartAsync(workflow);
```

In this example, `ReceiveHttpRequest` is an activity that triggers the workflow when an HTTP POST request is received at the specified path. The `ForEach` activity is used to iterate over a collection of data items, and for each item, it executes a sequence of activities (`ProcessDataItem` and `CustomActivity`). After processing all items, the `GenerateReport` activity generates a report, which is then sent back as an HTTP response using the `SendHttpResponse` activity.

Please note that this is a simplified example and assumes the existence of custom activities such as `ProcessDataItem`, `CustomActivity`, and `GenerateReport`. You would need to implement these activities based on your specific requirements.

For more information on how to create and use activities in Elsa Workflows, please refer to the official documentation:
- Programmatic workflows vs designer workflows: https://v3.elsaworkflows.io/docs/core-concepts/programmatic-vs-designer
- HTTP workflows: https://v3.elsaworkflows.io/docs/guides/http-workflows
- ForEach Loop Activity: https://stackoverflow.com/questions/66432903/using-a-elsa-workflow-foreach-loop-activity

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).
